homeassistant/components/binary_sensor/axis.py

```diff
@@ -20,48 +20,70 @@
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
     """Set up the Axis binary devices."""
-    add_entities([AxisBinarySensor(hass, discovery_info)], True)
+    add_entities([AxisBinarySensor(discovery_info)], True)
 
 
-class AxisBinarySensor(AxisDeviceEvent, BinarySensorDevice):
+class AxisBinarySensor(BinarySensorDevice):
     """Representation of a binary Axis event."""
 
-    def __init__(self, hass, event_config):
+    def __init__(self, event_config):
         """Initialize the Axis binary sensor."""
-        self.hass = hass
-        self._state = False
-        self._delay = event_config[CONF_TRIGGER_TIME]
-        self._timer = None
-        AxisDeviceEvent.__init__(self, event_config)
+        self.axis_event = event_config[CONF_EVENT]
+        self.device_name = event_config[CONF_NAME]
+        self.location = event_config[ATTR_LOCATION]
+        self.delay = event_config[CONF_TRIGGER_TIME]
+        self.remove_timer = None
 
-    @property
-    def is_on(self):
-        """Return true if event is active."""
-        return self._state
-
-    def update(self):
-        """Get the latest data and update the state."""
-        self._state = self.axis_event.is_tripped
+    async def async_added_to_hass(self):
+        """Subscribe sensors events."""
+        self.axis_event.callback = self._update_callback
 
     def _update_callback(self):
         """Update the sensor's state, if needed."""
-        self.update()
-
-        if self._timer is not None:
-            self._timer()
-            self._timer = None
+        if self.remove_timer is not None:
+            self.remove_timer()
+            self.remove_timer = None
 
-        if self._delay > 0 and not self.is_on:
-            # Set timer to wait until updating the state
+        if self.delay == 0 or self.is_on:
+            self.schedule_update_ha_state()
+        else:  # Run timer to delay updating the state
             def _delay_update(now):
                 """Timer callback for sensor update."""
                 _LOGGER.debug("%s called delayed (%s sec) update",
-                              self._name, self._delay)
+                              self.name, self.delay)
                 self.schedule_update_ha_state()
-                self._timer = None
+                self.remove_timer = None
 
-            self._timer = track_point_in_utc_time(
+            self.remove_timer = async_track_point_in_utc_time(
```

Comment:
Is this called from an async context? We're using `schedule_update_ha_state` above so I'm think it's a non async context.

Reply:
Yes. There are both sync and async parts to this implementation. 

Webrequests are using the requests library since it requires digest and aiohttp doesn't support it yet. Only used during setup.

Events are handled on a RTSP connection based on asyncio callback implementation. Which is everything related to sensors right now.

I'm just revisiting this for the first time in a year so am not up to speed on all parts yet myself.

Reply:
Changed it to be in async context.